### PR TITLE
[NOD-652] Add selected tip and get selected tip messages

### PR DESCRIPTION
**Pull request by someone235**

### CODE REVIEW

File: blockdag/dag.go [hunk 1]
```diff
@@ -1278,6 +1278,13 @@ func (dag *BlockDAG) isCurrent() bool {
 	return dagTimestamp >= minus24Hours
 }
 
+// AdjustedTime returns the adjusted time according to
+// dag.timeSource. See MedianTimeSource.AdjustedTime for
+// more details.
+func (dag *BlockDAG) AdjustedTime() time.Time {
+	return dag.timeSource.AdjustedTime()
```
Comment (svarogg) [inline blockdag/dag.go]: Find all the places where we call `dag.timeSource.AdjustedTime`, and replace with call to `dag.AdjustedTime()`

File: netsync/manager.go [hunk 2]
```diff
@@ -191,9 +199,26 @@ func (sm *SyncManager) startSync() {
 
 		syncPeer.PushGetBlockLocatorMsg(syncPeer.SelectedTip(), sm.dagParams.GenesisHash)
 		sm.syncPeer = syncPeer
-	} else {
-		log.Warnf("No sync peer candidates available")
+		return
+	}
+
+	log.Warnf("No sync peer candidates available")
+	if sm.dag.AdjustedTime().Sub(sm.dag.CalcPastMedianTime()) > time.Minute {
```
Comment (svarogg) [inline netsync/manager.go]: Extract this calculation to some function, the sending of selectedTip to another function and time.Minute to a constant.

File: netsync/manager.go [hunk 3]
```diff
@@ -191,9 +199,26 @@ func (sm *SyncManager) startSync() {
...
+		for peer, state := range sm.peerStates {
+			if !state.syncCandidate {
```
Comment (svarogg) [inline netsync/manager.go]: Can't he become a sync candidate once you asked for it's tips?

File: peer/peer.go [hunk 4]
```diff
@@ -644,10 +648,16 @@ func (p *Peer) SelectedTip() *daghash.Hash {
 	return selectedTip
 }
 
-// IsSyncCandidate returns whether or not this peer is a sync candidate.
+// SetSelectedTip sets the selected tip of the peer.
+func (p *Peer) SetSelectedTip(selectedTip *daghash.Hash) {
```
Comment (svarogg) [inline peer/peer.go]: I think it's a good idea that anywhere we have a selectedTip as hash - call it selectedTipHash.

Comment (svarogg) [reply peer/peer.go]: One exception I can think of - MsgSelectedTip and MsgGetSelectedTip can be without the Hash ending.

File: netsync/manager.go (hunk 3 above)
Comment (someone235) [reply netsync/manager.go]: you need to know its selected tip before you start syncing from it

File: netsync/manager.go [hunk 5]
```diff
@@ -195,32 +197,40 @@ func (sm *SyncManager) startSync() {
...
-			state.isPendingForSelectedTip = true
-			peer.QueueMessage(wire.NewMsgGetSelectedTip(), nil)
+			queueMsgGetSelectedTip(peer, state)
 		}
 	}
 	return
 }
 
+func (sm *SyncManager) shouldQueryPeerSelectedTips() bool {
+	return sm.dag.AdjustedTime().Sub(sm.dag.CalcPastMedianTime()) > time.Minute
```
Comment (svarogg) [inline netsync/manager.go]: I think we want minGetSelectedTipInterval here too

File: peer/peer.go [hunk 6]
```diff
@@ -168,6 +168,10 @@ type MessageListeners struct {
 	// message.
 	OnSendHeaders func(p *Peer, msg *wire.MsgSendHeaders)
 
+	OnGetSelectedTip func()
```
Comment (stasatdaglabs) [inline peer/peer.go]: Please add a comment above OnGetSelectedTip and OnSelectedTip.

File: peer/peer.go [hunk 7]
```diff
@@ -186,11 +190,11 @@ type MessageListeners struct {
 
 // Config is the struct to hold configuration options useful to Peer.
 type Config struct {
-	// SelectedTip specifies a callback which provides the selected tip
+	// SelectedTipHash specifies a callback which provides the selected tip
 	// to the peer as needed.
 	SelectedTip func() *daghash.Hash
```
Comment (stasatdaglabs) [inline peer/peer.go]: Rename this to SelectedTipHash.

File: peer/peer.go [hunk 8]
```diff
@@ -186,11 +190,11 @@ type MessageListeners struct {
...
 
-	// SelectedTip specifies a callback which provides the selected tip
+	// SelectedTipHashAndBlueScore specifies a callback which provides the selected tip
 	// to the peer as needed.
 	BlockExists func(*daghash.Hash) bool
```
Comment (stasatdaglabs) [inline peer/peer.go]: Rename this to SelectedTipHashAndBlueScore.

File: peer/peer.go [hunk 9]
```diff
@@ -486,7 +490,7 @@ func (p *Peer) StatsSnapshot() *StatsSnap {
 		TimeOffset:     p.timeOffset,
 		Version:        protocolVersion,
 		Inbound:        p.inbound,
-		SelectedTip:    p.selectedTip,
+		SelectedTip:    p.selectedTipHash,
```
Comment (stasatdaglabs) [inline peer/peer.go]: Rename this to SelectedTipHash.

File: peer/peer.go [hunk 10]
```diff
@@ -633,22 +637,28 @@ func (p *Peer) ProtocolVersion() uint32 {
...
 
-// SelectedTip returns the selected tip of the peer.
+// SelectedTipHash returns the selected tip of the peer.
 //
 // This function is safe for concurrent access.
-func (p *Peer) SelectedTip() *daghash.Hash {
+func (p *Peer) SelectedTipHash() *daghash.Hash {
 	p.statsMtx.RLock()
-	selectedTip := p.selectedTip
+	selectedTip := p.selectedTipHash
```
Comment (stasatdaglabs) [inline peer/peer.go]: Rename this to selectedTipHash.

File: server/p2p/on_selected_tip.go [hunk 11]
```diff
@@ -0,0 +1,14 @@
+package p2p
+
+import (
+	"github.com/kaspanet/kaspad/peer"
+	"github.com/kaspanet/kaspad/wire"
+)
+
+// OnSelectedTip is invoked when a peer receives a selectedTipHash kaspa
```
Comment (stasatdaglabs) [inline server/p2p/on_selected_tip.go]: selectedTipHash -> selectedTip

File: server/p2p/p2p.go [hunk 12]
```diff
@@ -1046,10 +1046,12 @@ func newPeerConfig(sp *Peer) *peer.Config {
 			OnFilterLoad:      sp.OnFilterLoad,
 			OnGetAddr:         sp.OnGetAddr,
 			OnAddr:            sp.OnAddr,
+			OnGetSelectedTip:  sp.OnGetSelectedTip,
+			OnSelectedTip:     sp.OnSelectedTip,
 			OnRead:            sp.OnRead,
 			OnWrite:           sp.OnWrite,
 		},
-		SelectedTip:       sp.selectedTip,
+		SelectedTip:       sp.selectedTipHash,
```
Comment (stasatdaglabs) [inline server/p2p/p2p.go]: Rename this to SelectedTipHash.

File: wire/msggetselectedtip.go [hunk 13]
```diff
@@ -0,0 +1,41 @@
+package wire
+
+import (
+	"io"
+)
+
+// MsgGetSelectedTip implements the Message interface and represents a kaspa
+// getseltip message. It is used to request the selected tip of another tip.
```
Comment (stasatdaglabs) [inline wire/msggetselectedtip.go]: another tip -> another peer

File: wire/msgselectedtip.go [hunk 14]
```diff
@@ -0,0 +1,62 @@
+// Copyright (c) 2013-2016 The btcsuite developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
```
Comment (stasatdaglabs) [inline wire/msgselectedtip.go]: Please remove, the copyright notice in the root folder should suffice.

File: wire/msgselectedtip.go [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+package wire
+
+import (
+	"github.com/kaspanet/kaspad/util/daghash"
+	"io"
+)
+
+// MsgSelectedTip implements the Message interface and represents a kaspa
+// selectedtip message. It is used to answer getseltip messages and tell
+// the asking peer what is your selected tip.
```
Comment (stasatdaglabs) [inline wire/msgselectedtip.go]: your -> our/this peer's

File: wire/msgselectedtip.go [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+type MsgSelectedTip struct {
+	// The selected tip hash of the generator of the message.
+	SelectedTipHash *daghash.Hash
+}
+
+// KaspaDecode decodes r using the kaspa protocol encoding into the receiver.
+// The version message is special in that the protocol version hasn't been
```
Comment (stasatdaglabs) [inline wire/msgselectedtip.go]: Why does this comment speak about the version message?

File: wire/msgselectedtip_test.go [hunk 17]
```diff
@@ -0,0 +1,94 @@
+// Copyright (c) 2013-2016 The btcsuite developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
```
Comment (stasatdaglabs) [inline wire/msgselectedtip_test.go]: Please remove, the copyright notice in the root folder should suffice.

File: wire/msgversion.go [hunk 18]
```diff
@@ -56,7 +56,7 @@ type MsgVersion struct {
 	UserAgent string
 
 	// The selected tip of the generator of the version message.
```
Comment (stasatdaglabs) [inline wire/msgversion.go]: The selected tip **hash**

File: wire/msgversion_test.go [hunk 19]
```diff
@@ -55,9 +55,9 @@ func TestVersion(t *testing.T) {
 		t.Errorf("NewMsgVersion: wrong user agent - got %v, want %v",
 			msg.UserAgent, DefaultUserAgent)
 	}
-	if !msg.SelectedTip.IsEqual(selectedTip) {
+	if !msg.SelectedTipHash.IsEqual(selectedTip) {
 		t.Errorf("NewMsgVersion: wrong selected tip - got %s, want %s",
```
Comment (stasatdaglabs) [inline wire/msgversion_test.go]: wrong selected tip **hash**

File: peer/peer_test.go [hunk 20]
```diff
@@ -578,7 +578,7 @@ func TestOutboundPeer(t *testing.T) {
 		return hash
 	}
 
-	peerCfg.SelectedTip = selectedTip
+	peerCfg.SelectedTipHash = selectedTip
```
Comment (stasatdaglabs) [inline peer/peer_test.go]: selectedTip -> selectedTipHash

File: server/p2p/p2p.go [hunk 21]
```diff
@@ -296,8 +296,8 @@ func newServerPeer(s *Server, isPersistent bool) *Peer {
 	}
 }
 
-// selectedTip returns the current selected tip
-func (sp *Peer) selectedTip() *daghash.Hash {
+// selectedTipHash returns the current selected tip
```
Comment (stasatdaglabs) [inline server/p2p/p2p.go]: selected tip **hash**

**Pull request merged**